librsvg problem in clang

### Description / Steps to reproduce the issue

building librsvg with clang produces this error ->

```
ld.lld: error: librsvg_c_api.a(bcryptprimitives.dll): .idata$4 should not refer to special section 0
clang: error: linker command failed with exit code 1 (use -v to see invocation)

```


### Expected behavior

should produce the librsvg build like with gcc

### Actual behavior

bugs out with the above error

### Verification

- [X] I have verified that my MSYS2 is up-to-date before submitting the report (see https://www.msys2.org/docs/updating/)

### Windows Version

MSYS_NT-10.0-26100

### MINGW environments affected

- [ ] MINGW64
- [ ] MINGW32
- [ ] UCRT64
- [X] CLANG64
- [X] CLANG32
- [ ] CLANGARM64

### Are you willing to submit a PR?

_No response_

Thank you so much for this. So the problem might lie within the librsvg build system because IIUC it's not Rust who added that `--whole-archive`.

Question to anyone who knows autotools better:
Is this `--whole-archive` added by `libtool`?
```
$ LANG=C make -j32 V=1
make  all-recursive
make[1]: Entering directory '/h/projects/librsvg'
Making all in .
make[2]: Entering directory '/h/projects/librsvg'
/bin/sh ./libtool  --tag=CC   --mode=link cc  -g -O2  -version-info 52:0:50 -export-dynamic -no-undefined -export-symbols-regex "^rsvg_.*" "-Wl,--gc-sections"    -o librsvg-2.la -rpath /clang64/lib   librsvg_c_api.la -lpng16 -lcairo-gobject -lfreetype -lgdk_pixbuf-2.0 -lgio-2.0 -lxml2 -lpangocairo-1.0 -lpango-1.0 -lgobject-2.0 -lglib-2.0 -lintl -lharfbuzz -lcairo -lm  -lws2_32 -luserenv -lbcrypt -lntdll -v
libtool: link: rm -fr  .libs/librsvg-2.exp
libtool: link: /clang64/bin/nm -B   ./.libs/librsvg_c_api.a | /usr/bin/sed -n -e 's/^.*[         ]\([ABCDGIRSTW][ABCDGIRSTW]*\)[         ][      ]*\([_A-Za-z][_A-Za-z0-9]*\)\{0,1\}$/\1 \2 \2/p' | /usr/bin/sed '/ __gnu_lto/d' | /usr/bin/sed -e '/^[BCDGRS][ ]/s/.*[ ]\([^ ]*\)/\1 DATA/;s/^.*[ ]__nm__\([^ ]*\)[ ][^ ]*/\1 DATA/;/^I[ ]/d;/^[AITW][ ]/s/.* //' | sort | uniq > .libs/librsvg-2.exp
libtool: link: /usr/bin/grep -E -e "^rsvg_.*" ".libs/librsvg-2.exp" > ".libs/librsvg-2.expT"
libtool: link: mv -f ".libs/librsvg-2.expT" ".libs/librsvg-2.exp"
libtool: link: if test DEF = "`/usr/bin/sed -n -e 's/^[  ]*//' -e '/^\(;.*\)*$/d' -e 's/^\(EXPORTS\|LIBRARY\)\([         ].*\)*$/DEF/p' -e q .libs/librsvg-2.exp`" ; then cp .libs/librsvg-2.exp .libs/librsvg-2-2.dll.def; else echo EXPORTS > .libs/librsvg-2-2.dll.def; cat .libs/librsvg-2.exp >> .libs/librsvg-2-2.dll.def; fi
libtool: link:  cc -shared .libs/librsvg-2-2.dll.def  -Wl,--whole-archive ./.libs/librsvg_c_api.a -Wl,--no-whole-archive  -lpng16 -lcairo-gobject -lfreetype -lgdk_pixbuf-2.0 -lgio-2.0 -lxml2 -lpangocairo-1.0 -lpango-1.0 -lgobject-2.0 -lglib-2.0 -lintl -lharfbuzz -lcairo -lws2_32 -luserenv -lbcrypt -lntdll  -g -O2 -Wl,--gc-sections   -o .libs/librsvg-2-2.dll -Wl,--enable-auto-image-base -Xlinker --out-implib -Xlinker .libs/librsvg-2.dll.a
ld.lld: error: librsvg_c_api.a(bcryptprimitives.dll): .idata$4 should not refer to special section 0
cc: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [Makefile:900: librsvg-2.la] Error 1
make[2]: Leaving directory '/h/projects/librsvg'
make[1]: *** [Makefile:1158: all-recursive] Error 1
make[1]: Leaving directory '/h/projects/librsvg'
make: *** [Makefile:793: all] Error 2
```
Talking about this line: `libtool: link:  cc -shared .libs/librsvg-2-2.dll.def  -Wl,--whole-archive ./.libs/librsvg_c_api.a ....`, I have added `-v` to line 900, as you can see in the log above, but it's not the failing command.
[Makefile.txt](https://github.com/user-attachments/files/15903919/Makefile.txt)

--- a/lld/COFF/InputFiles.cpp
+++ b/lld/COFF/InputFiles.cpp
@@ -746,6 +746,26 @@ std::optional<Symbol *> ObjFile::createDefined(
   if (sectionNumber == llvm::COFF::IMAGE_SYM_DEBUG)
     return nullptr;
 
+  if (sym.isEmptySectionDeclaration()) {
+    // As there is no coff_section in the object file for these, make a
+    // new virtual one, with everything zeroed out (i.e. an empty section),
+    // with only the name and characteristics set.
+    StringRef name = getName();
+    auto *hdr = make<coff_section>();
+    memset(hdr, 0, sizeof(*hdr));
+    strncpy(hdr->Name, name.data(),
+            std::min(name.size(), (size_t)COFF::NameSize));
+    // We have no idea what characteristics should be assumed here; pick
+    // a default. This matches what is used for .idata sections in the regular
+    // object files in import libraries.
+    hdr->Characteristics = IMAGE_SCN_CNT_INITIALIZED_DATA | IMAGE_SCN_MEM_READ |
+                           IMAGE_SCN_MEM_WRITE | IMAGE_SCN_ALIGN_4BYTES;
+    auto *sc = make<SectionChunk>(this, hdr);
+    chunks.push_back(sc);
+    return make<DefinedRegular>(this, /*name=*/"", /*isCOMDAT=*/false,
+                                /*isExternal=*/false, sym.getGeneric(), sc);
+  }
+
   if (llvm::COFF::isReservedSectionNumber(sectionNumber))
     Fatal(ctx) << toString(this) << ": " << getName()
                << " should not refer to special section "

--- a/lld/test/COFF/empty-section-decl.yaml
+++ b/lld/test/COFF/empty-section-decl.yaml
@@ -1,0 +1,56 @@
+# REQUIRES: x86
+
+# RUN: yaml2obj %s -o %t.obj
+# RUN: lld-link -dll -out:%t.dll %t.obj -noentry -subsystem:console -lldmap:%t.map
+# RUN: llvm-objdump -s %t.dll | FileCheck %s
+# RUN: FileCheck %s --check-prefix=MAP < %t.map
+
+# CHECK:      Contents of section .itest:
+# CHECK-NEXT:  180001000 0c100080 01000000 00000000 01000000
+
+# MAP: 00001000 0000000a     4         {{.*}}:(.itest$2)
+# MAP: 00001000 00000000     0                 .itest$2
+# MAP: 0000100c 00000000     4         {{.*}}:(.itest$4)
+# MAP: 0000100c 00000000     0                 .itest$4
+# MAP: 0000100c 00000004     2         {{.*}}:(.itest$6)
+# MAP: 0000100c 00000000     0                 .itest$6
+
+--- !COFF
+header:
+  Machine:         IMAGE_FILE_MACHINE_AMD64
+  Characteristics: [  ]
+sections:
+  - Name:            '.itest$2'
+    Characteristics: [ IMAGE_SCN_CNT_INITIALIZED_DATA, IMAGE_SCN_MEM_READ, IMAGE_SCN_MEM_WRITE ]
+    Alignment:       4
+    SectionData:     '00000000000000000000'
+    SizeOfRawData:   10
+    Relocations:
+      - VirtualAddress:  0
+        SymbolName:      '.itest$4'
+        Type:            IMAGE_REL_AMD64_ADDR64
+  - Name:            '.itest$6'
+    Characteristics: [ IMAGE_SCN_CNT_INITIALIZED_DATA, IMAGE_SCN_MEM_READ, IMAGE_SCN_MEM_WRITE ]
+    Alignment:       2
+    SectionData:     01000000
+    SizeOfRawData:   4
+symbols:
+  - Name:            '.itest$2'
+    Value:           0
+    SectionNumber:   1
+    SimpleType:      IMAGE_SYM_TYPE_NULL
+    ComplexType:     IMAGE_SYM_DTYPE_NULL
+    StorageClass:    IMAGE_SYM_CLASS_SECTION
+  - Name:            '.itest$6'
+    Value:           0
+    SectionNumber:   2
+    SimpleType:      IMAGE_SYM_TYPE_NULL
+    ComplexType:     IMAGE_SYM_DTYPE_NULL
+    StorageClass:    IMAGE_SYM_CLASS_STATIC
+  - Name:            '.itest$4'
+    Value:           0
+    SectionNumber:   0
+    SimpleType:      IMAGE_SYM_TYPE_NULL
+    ComplexType:     IMAGE_SYM_DTYPE_NULL
+    StorageClass:    IMAGE_SYM_CLASS_SECTION
+...

--- a/lld/test/COFF/wholearchive-implib.s
+++ b/lld/test/COFF/wholearchive-implib.s
@@ -1,0 +1,35 @@
+// REQUIRES: x86
+// RUN: split-file %s %t.dir
+// RUN: llvm-lib -machine:amd64 -out:%t.lib -def:%t.dir/lib.def
+// RUN: llvm-mc -filetype=obj -triple=x86_64-windows %t.dir/main.s -o %t.main.obj
+
+// RUN: lld-link -out:%t.exe %t.main.obj -wholearchive:%t.lib -entry:entry -subsystem:console
+// RUN: llvm-readobj --coff-imports %t.exe | FileCheck %s
+
+// As LLD usually doesn't use the header/trailer object files from import
+// libraries, but instead synthesizes those structures, we end up with two
+// import directory entries if we force those objects to be included.
+
+// CHECK:      Import {
+// CHECK-NEXT:   Name: lib.dll
+// CHECK-NEXT:   ImportLookupTableRVA: 0x2050
+// CHECK-NEXT:   ImportAddressTableRVA: 0x2068
+// CHECK-NEXT: }
+// CHECK-NEXT: Import {
+// CHECK-NEXT:   Name: lib.dll
+// CHECK-NEXT:   ImportLookupTableRVA: 0x2058
+// CHECK-NEXT:   ImportAddressTableRVA: 0x2070
+// CHECK-NEXT:   Symbol: func (0)
+// CHECK-NEXT: }
+
+
+#--- main.s
+.global entry
+entry:
+  call func
+  ret
+
+#--- lib.def
+LIBRARY lib.dll
+EXPORTS
+func

--- a/llvm/include/llvm/Object/COFF.h
+++ b/llvm/include/llvm/Object/COFF.h
@@ -392,6 +392,11 @@ public:
            getValue() == 0;
   }
 
+  bool isEmptySectionDeclaration() const {
+    return isSection() && getSectionNumber() == COFF::IMAGE_SYM_UNDEFINED &&
+           getValue() == 0;
+  }
+
   bool isWeakExternal() const {
     return getStorageClass() == COFF::IMAGE_SYM_CLASS_WEAK_EXTERNAL;
   }

